data-exporter/src/main/java/org/reactome/release/NCBIGene.java

```diff
@@ -0,0 +1,204 @@
+	 */
+	public void writeGeneXMLFiles(Session graphDBSession) throws IOException {
+
+		logger.info("Writing gene XML file(s)");
+
+		Path geneErrorFilePath = getGeneErrorFilePath();
+		deleteAndCreateFile(geneErrorFilePath);
+
+		Set<String> ncbiGeneXMLNodeStrings = new LinkedHashSet<>();
+		for (NCBIEntry ncbiEntry : ncbiEntries) {
+			ncbiGeneLogger.info("Working on " + ncbiEntry.getUniprotAccession());
+
+			Set<ReactomeEvent> topLevelPathways = ncbiEntry.getTopLevelPathways(graphDBSession);
+			if (topLevelPathways.isEmpty()) {
+				String errorMessage = ncbiEntry.getUniprotDisplayName() +
+									  " participates in Event(s) but no top Pathway can be found, i.e. there seem to be a pathway" +
+									  " which contains or is an instance of itself.\n";
+
+				Files.write(geneErrorFilePath, errorMessage.getBytes(), StandardOpenOption.APPEND);
+				continue;
+			}
+
+			for (String ncbiGeneId : ncbiEntry.getNcbiGeneIds()) {
+				ncbiGeneXMLNodeStrings.add(ncbiEntry.getEntityLinkXML(ncbiGeneId));
+
+				for (ReactomeEvent topLevelPathway : topLevelPathways) {
+					ncbiGeneXMLNodeStrings.add(ncbiEntry.getEventLinkXML(ncbiGeneId, topLevelPathway));
+				}
+			}
+
+			ncbiGeneLogger.info("Finished with " + ncbiEntry.getUniprotAccession());
+		}
+
+		int fileCount = 0;
+		int numberOfGeneXMLFiles = getNumberOfGeneXMLFiles(ncbiGeneXMLNodeStrings);
+		for (Set<String> ncbiGeneXMLNodeStringsSubSet : Utilities.splitSet(ncbiGeneXMLNodeStrings, numberOfGeneXMLFiles)) {
+			Path geneXMLFilePath = getGeneXMLFilePath(++fileCount);
+
+			deleteAndCreateFile(geneXMLFilePath);
+
+			logger.info("Generating " + geneXMLFilePath.getFileName());
+
+			appendWithNewLine(getXMLHeader(), geneXMLFilePath);
+			appendWithNewLine(getOpenRootTag(), geneXMLFilePath);
+			for (String ncbiGeneXMLNodeString : ncbiGeneXMLNodeStringsSubSet) {
+				appendWithNewLine(ncbiGeneXMLNodeString, geneXMLFilePath);
+			}
+
+			appendWithNewLine(getCloseRootTag(), geneXMLFilePath);
+		}
+
+		logger.info("Finished writing gene XML file(s)");
+	}
+
+	/**
+	 * The size (in bytes) for the set of all Link XML node strings to include in the final NCBI Gene XML files
+	 * is determined and the minimum number of files to write in order for each to be less than the pre-set limit is
+	 * returned.
+	 * The maximum size of each file is 15MB and is a limit set by NCBI for individual file uploads.
+	 * @param ncbiGeneXMLNodeStrings Set of all the Link XML nodes strings to include in the NCBI Gene XML file(s)
+	 * @return Number of NCBI Gene XML files to create (i.e. split the XML nodes amongst to be under the file size
+	 * limit)
+	 */
+	private int getNumberOfGeneXMLFiles(Set<String> ncbiGeneXMLNodeStrings) {
+		final int BYTES_TO_KILOBYTES = 1024;
+		final int KILOBYTES_TO_MEGABYTES = 1024;
+		final double MAX_FILE_SIZE_IN_MEGABYTES = 15.0;
+
+		int sizeInBytes = ncbiGeneXMLNodeStrings.stream().mapToInt(str -> str.getBytes().length).sum();
```

Comment:
I haven't tested this - is it an accurate reflection of how many bytes will be required to write this to disc? I would expect that different character sets/encodings might change the size reported here, so make sure you write with the same charset/encoding that this line uses.

 Or, just lower MAX_FILE_SIZE_IN_MEGABYTES to, I dunno, 13 or 14? for a little wiggle room. ;)

Reply:
That's a good idea.  In testing with release 67 data, it worked as expected but I was worried about an edge case of adding the XML header and root tag pushing it just over the edge of 15MB :).  A megabyte of wiggle room sounds good to me.